File: VoiceType/Services/DiarizationService.swift
import Foundation

enum DiarizationError: LocalizedError {
    case modelMissing

    var errorDescription: String? {
        switch self {
        case .modelMissing:
            return "说话人分离模型未安装，请运行 scripts/setup_diarization.sh"
        }
    }
}

struct SpeakerSegment: Codable, Equatable {
    var speaker: Int
    var start: Double
    var end: Double
    var text: String = ""
}

/// 说话人分离：pyannote 分段 + 3D-Speaker 声纹聚类（sherpa-onnx，纯本地）。
/// 推理在专用串行队列执行，模型懒加载常驻。
final class DiarizationService: @unchecked Sendable {
    private var wrapper: SherpaOnnxOfflineSpeakerDiarizationWrapper?
    private var loadedNumClusters: Int = .min
    private let queue = DispatchQueue(
        label: "com.catsonkeyboard.voicetype.diarization", qos: .userInitiated)

    /// numSpeakers 为 nil 时自动估计人数
    func diarize(samples: [Float], numSpeakers: Int?) async throws -> [SpeakerSegment] {
        try await withCheckedThrowingContinuation { cont in
            queue.async {
                do {
                    cont.resume(
                        returning: try self.diarizeSync(
                            samples: samples, numSpeakers: numSpeakers))
                } catch {
                    cont.resume(throwing: error)
                }
            }
        }
    }

    private func diarizeSync(samples: [Float], numSpeakers: Int?) throws -> [SpeakerSegment] {
        guard ModelPaths.diarizationPresent else { throw DiarizationError.modelMissing }
        let clusters = numSpeakers ?? -1
        if wrapper == nil || loadedNumClusters != clusters {
            var config = sherpaOnnxOfflineSpeakerDiarizationConfig(
                segmentation: sherpaOnnxOfflineSpeakerSegmentationModelConfig(
                    pyannote: sherpaOnnxOfflineSpeakerSegmentationPyannoteModelConfig(
                        model: ModelPaths.segmentationModel.path),
                    numThreads: 2),
                embedding: sherpaOnnxSpeakerEmbeddingExtractorConfig(
                    model: ModelPaths.speakerEmbeddingModel.path, numThreads: 2),
                clustering: sherpaOnnxFastClusteringConfig(
                    numClusters: clusters, threshold: 0.5))
            wrapper = SherpaOnnxOfflineSpeakerDiarizationWrapper(config: &config)
            loadedNumClusters = clusters
        }
        guard let wrapper, wrapper.impl != nil else { throw DiarizationError.modelMissing }
        return wrapper.process(samples: samples).map {
            SpeakerSegment(speaker: $0.speaker, start: Double($0.start), end: Double($0.end))
        }
    }

    /// 相邻同说话人且间隔 ≤ maxGap 秒的段合并（减少转写碎片）
    static func mergeAdjacent(_ segments: [SpeakerSegment], maxGap: Double = 1.0)
        -> [SpeakerSegment]
    {
        var result: [SpeakerSegment] = []
        for seg in segments.sorted(by: { $0.start < $1.start }) {
            if var last = result.last, last.speaker == seg.speaker,
                seg.start - last.end <= maxGap
            {
                last.end = max(last.end, seg.end)
                result[result.count - 1] = last
            } else {
                result.append(seg)
            }
        }
        return result
    }
}

File: windows/src/VoiceType/Services/DiarizationService.cs
using VoiceType.Interop.SherpaOnnx;
using VoiceType.Models;

namespace VoiceType.Services;

public sealed class DiarizationException : Exception
{
    public DiarizationException(string message) : base(message) { }
}

/// <summary>
/// 说话人分离：pyannote 分段 + 3D-Speaker 声纹聚类（sherpa-onnx，纯本地）。
/// 推理在串行信号量下执行，模型懒加载常驻。
/// </summary>
public sealed class DiarizationService : IDisposable
{
    private readonly SemaphoreSlim _gate = new(1, 1);
    private OfflineSpeakerDiarization? _wrapper;
    private int _loadedNumClusters = int.MinValue;

    /// <summary>numSpeakers 为 null 时自动估计人数。推理串行 + 线程池执行。</summary>
    public async Task<List<SpeakerSegment>> DiarizeAsync(
        float[] samples, int? numSpeakers, CancellationToken ct = default)
    {
        await _gate.WaitAsync(ct);
        try
        {
            return await Task.Run(() => DiarizeSync(samples, numSpeakers), ct);
        }
        finally
        {
            _gate.Release();
        }
    }

    private List<SpeakerSegment> DiarizeSync(float[] samples, int? numSpeakers)
    {
        if (!ModelPaths.DiarizationPresent)
            throw new DiarizationException(
                "说话人分离模型未安装，请运行 windows/scripts/setup_diarization.ps1");
        int clusters = numSpeakers ?? -1;
        if (_wrapper is null || _loadedNumClusters != clusters)
        {
            _wrapper?.Dispose();
            _wrapper = OfflineSpeakerDiarization.Create(new DiarizationOptions
            {
                SegmentationModel = ModelPaths.SegmentationModel,
                EmbeddingModel = ModelPaths.SpeakerEmbeddingModel,
                SegmentationNumThreads = 2,
                EmbeddingNumThreads = 2,
                NumClusters = numSpeakers,
            });
            _loadedNumClusters = clusters;
        }
        return _wrapper.Process(samples)
            .Select(s => new SpeakerSegment { Speaker = s.Speaker, Start = s.Start, End = s.End })
            .ToList();
    }

    /// <summary>相邻同说话人且间隔 ≤ maxGap 秒的段合并（减少转写碎片）</summary>
    public static List<SpeakerSegment> MergeAdjacent(
        List<SpeakerSegment> segments, double maxGap = 1.0)
    {
        var result = new List<SpeakerSegment>();
        foreach (SpeakerSegment seg in segments.OrderBy(s => s.Start))
        {
            if (result.Count > 0 &&
                result[^1].Speaker == seg.Speaker &&
                seg.Start - result[^1].End <= maxGap)
            {
                result[^1].End = Math.Max(result[^1].End, seg.End);
            }
            else
            {
                result.Add(new SpeakerSegment
                {
                    Speaker = seg.Speaker, Start = seg.Start, End = seg.End, Text = seg.Text,
                });
            }
        }
        return result;
    }

    public void Dispose()
    {
        _wrapper?.Dispose();
        _wrapper = null;
        _gate.Dispose();
    }
}
